Multiple Element Growth Options && Spirit Image autosizing & scaling

resonant-gamedesign opened the pull request:
Two changes:

For Element Growth, this implements OR and 2 of the same (AND not implemented)
Required adding a 'backslash' image to separate the OR elements
Required a change that may have consequences - growth-cell gain icon. is now "display: inline-block;" instead of just block
Also, I changed the bottom margin on the elements to 7px from 10px... this made them line up better.

Changing the image sizing required reworking the order of things that happen in MAIN (so that the special rule can resize before the image sizes itself).

I hope I'm doing things correctly. Never used GitHub before.

resonant-gamedesign committed:
Update to element growth options

Update allows 2 element growth options (Downpour) and OR element growth options (Lure).

For the 2 element option:
icon-gtop icon
icon-gbottom icon

For the OR element option:
growth-cell gain icon.orelement

Additionally, the bottom margin on growth-cell icon.X was changed to 7px because it lines up better. There may be a better solution.

resonant-gamedesign committed:
Update board_front.js

Update includes two main changes. 
First, it allows for multiple elements in growth (OR or multiple of the same). It does not allow for AND, which isn't' currently used in the game but is a reasonable design choice.
Second, it updates how the picture is scaled. The picture now automatically scales to the size of the window (accounting for the changing Special Rule size). The user can change the image size using the spirit-image-size option that was already in the HTML (but not used, to my knowledge).
Requires a change to Growth options.css as well

resonant-gamedesign committed:
Backslash image (for OR Element Growth)

I created this image for use as a backslash between OR element growth options (think Lure).
This is required for my other Pull request on the board.js

resonant-gamedesign committed:
Merge pull request #2 from resonant-gamedesign/resonant-gamedesign-patch-2

Backslash image (for OR Element Growth)

resonant-gamedesign committed:
Merge pull request #1 from resonant-gamedesign/resonant-gamedesign-patch-1

Update board_front.js

LokiMidgard commented:
This looks correct, for a pull request ;)

This is not required but especially when changing somthing graphically I like to put screenshots/images of those changes in the pull request, or at least some examples.

But I also havn't done it in every pull request I did...

resonant-gamedesign committed:
A higher-res escalation icon

scaled down to 25px from the escalation icon on the wiki

resonant-gamedesign committed:
Update adversary.css

Adds "Base Difficulty" to the CSS
Adds "Difficulty Level" to the CSS (that's the little number next to each level)
Requires changes to HTML - this seems to only be available in the example. Should we create a folder for the html with no content? I also think an empty folder for folk to add their designs to would be helpful. I had trouble getting started initially.

Updates to the following icons:
Escalation
City
Presence

resonant-gamedesign committed:
Update adversary.html

Adds the 'Base Difficulty' text, adds difficulty to the levels, and removes 'Stage II' from 'Stage II Escalation' to match how Russia and Scotland appear (which I assume is Eric's latest design).
Requires changes to CSS

resonant-gamedesign commented:
Are the three commits related to the adversary already part of the pull?
Will this pull update with every change I make to my master? If that's the case, I guess I should branch my master?
I'll add images next time - where do I add those? As comments?

resonant-gamedesign commented:
![Example](https://user-images.githubusercontent.com/44596045/151403412-bdc9aa34-bee9-4df5-ac8f-95a97bae50e0.PNG)
See example for 1) spirit image scaling (though you'll want to play around with that), 2) an example of 2 of the same element, and 3) an example of an OR (it works for 3 elements as well, and more but once it overflows bad things happen).

LokiMidgard commented:
> Are the three commits related to the adversary already part of the pull? Will this pull update with every change I make to my master? If that's the case, I guess I should branch my master? I'll add images next time - where do I add those? As comments?

Yes from the branch you created the pull request you get all new commits. If you want to remove some commits you could force push to your branch I think...

You can also edit the original pull request if you want.

resonant-gamedesign commented:
I'm OK with these all being one big push, but I'll segment future work so its not overwhelming. I assume whoever is moderating the master file will want changes to be clear and separate instead of a whole mess of them at once.

resonant-gamedesign commented on _examples/adversary/adversary.html:
```diff
@@ -16,7 +17,7 @@
       </loss-condition>
       <escalation>
-        <section-title>Stage II Escalation <icon class="escalation"></icon></section-title>
+        <section-title>Escalation <icon class="escalation"></icon></section-title>
```
Russia and Scotland dropped 'Stage II'. This updates the template to match.

resonant-gamedesign commented on _global/css/adversary.css:
```diff
@@ -59,6 +59,21 @@ adversary-title{
+  font-family: 'DK Snemand';
+  text-transform: uppercase;
+}
+
```
This positions the Base Difficulty text. I didn't really put much thought into the height/width so if folks see any concerns there, please edit.

resonant-gamedesign commented on _global/css/adversary.css:
```diff
@@ -195,6 +214,19 @@ icon.city{
+}
+
+icon.presence{
+  background-image: url('../images/board/50px_presence.png');
```
I updated some of the blurry icons to use higher resolution icons. I think all of these already exist in the master repo, but I'll have to double check. If the icons disappear when this is merged, that's why.

Yexo replied:
Let's check this before merging.

resonant-gamedesign replied:
I've added some icons since - simple elements and other things. Thoughts/concerns? Is file size an issue? image folder is 15MB

resonant-gamedesign commented on _global/js/board_front.js:
```diff
@@ -39,13 +42,24 @@ function addImages(board) {
     if(spiritImage){
-        board.innerHTML = `<div class="spirit-image" style="background-image: url(${spiritImage});" ></div>` + board.innerHTML
+		//Image now scales to fill gap. 'imageSize' allows the user to specify what % of the gap to cover
+        board.innerHTML = `<div class="spirit-image" style="background-image: url(${spiritImage}); background-size: auto ${imageSize}; height:calc(100% - ${height}); width:1700px;" ></div>` + board.innerHTML
```
The 'width:1700px' here is arbitrary and I'm open to other fixes. I'd like the width to be unbounded but I wasn't able to find code that would do so and still work with the image scaling.

resonant-gamedesign commented on _global/js/board_front.js:
```diff
@@ -138,6 +152,11 @@ function parseGrowthTags(){
+				case 'reclaim-none':
+                    {
+                        newGrowthCellHTML += `${openTag}{reclaim-none}<growth-text>Reclaim None</growth-text></growth-cell>`
+                        break;
```
I shouldn't have committed this change - its a joke icon for my Water Acts Not spirit.
Can you reject specific lines?

Yexo replied:
No. You can either go back in your tree and modify the old commit, then force-push to your own branch. Alternatively (and this is easier) just add another commit on top removing these lines again.

resonant-gamedesign commented on _global/css/board_front/growth-options.css:
```diff
@@ -214,6 +214,10 @@ growth-cell icon.reclaim-one {
+growth-cell icon.reclaim-none {
+  margin-bottom: 7px;
+}
+
```
This is a joke icon from my Water Acts Not spirit. We don't need to merge this particular change (not sure if you can accept changes line-by-line).

resonant-gamedesign commented on _global/css/board_front/growth-options.css:
```diff
@@ -389,12 +391,19 @@ growth-cell gain icon.sun,
   background-repeat: no-repeat;
   background-size: contain;
-  display: block;
+  display: inline-block;
```
I found this to be required to make the elements appear adjacent to one another instead of on top of one another. I'm not sure if this will later cause issues.

resonant-gamedesign commented on _global/css/board_front/growth-options.css:
```diff
@@ -410,5 +419,16 @@ growth-cell icon.moon {
-  margin-bottom: 10px;
-}
\ No newline at end of file
+  margin-bottom: 7px;
```
I found 10px was too many. 7px lines up better.

resonant-gamedesign commented on _global/css/board_front/growth-options.css:
```diff
@@ -410,5 +419,16 @@ growth-cell icon.moon {
+icon-gbottom icon {
+  transform: translateY(0px) translateX(-14px);
+  display: inline-block;
+}
```
Old edit - now improved with a more dynamic element image placement

resonant-gamedesign committed:
simple_element_innates

This commit changes innate powers to use the simple element icons instead of the detailed ones (this matches official content).

resonant-gamedesign committed:
special rule icon size

In special rules, updated the icon sizes for sacred sites and cities.

resonant-gamedesign committed:
Commenting on growth options

Just comments

resonant-gamedesign committed:
Icon updates

Cleaned up some icons with more recent version from the wiki or other small tweaks to improve appearance.

resonant-gamedesign committed:
Fix default target icon size

This change fixes the issue that having a target with an OR condition would overflow the target box.

resonant-gamedesign committed:
higher resolution sacred site

updating the sacred site icon to default to a higher resolution.

resonant-gamedesign committed:
Threshold icons moved

Moved these threshold icons to innate powers.
Will need to check how thresholds are handled in power cards...

resonant-gamedesign committed:
various new icons

adding various icons, including 'reclaim none' which is really just a joke.

resonant-gamedesign committed:
adding css for the backslash icon + other

the other changes related to the destroyed presence icon (think Volcano) and the margins on the normal presence icon (so it falls in line better).

resonant-gamedesign commented:
I think I'm doing something wrong. I don't mean to push all my changes to the main repository, and certainly not all at once. Any clever solutions?

resonant-gamedesign committed:
Clarifications

I'm trying to clear up some of the start-up instructions. I had trouble with my initial creation until I figure out that I should put it in a folder.

resonant-gamedesign committed:
Use simple elements in special rules

update changes special rules to use simple elements.
Also updates dahan in special rules to make them line up a bit better. Line spacing is odd.

resonant-gamedesign committed:
Single line - centered

The single line innate power should be centered (ie. middle) not on the bottom of the line. This fixes that, sort of.

resonant-gamedesign committed:
Update innate-power.css

resonant-gamedesign committed:
Updated destroyed presence icon

resonant-gamedesign committed:
Simple elements in Innate descriptions

This change updates innate descriptions to use the simple elements (matching official content)

resonant-gamedesign committed:
Icon specification in specific ccs sheets

The icon.css sheet was taking priority over more specific icon designations (for example, in special rules or the notes on innate powers). These changes allow the more specific rules to take priority.

resonant-gamedesign committed:
Undoing change to Destroyed Presence

This one is from the wiki

resonant-gamedesign committed:
allowing reverse-named dual lands

I don't expect users to remember if its wetland-jungle or jungle-wetland. This solves that problem.
However, it may cause some issues down the line if these icons are used by name in other places. I've tried to find a way to make one class be interpreted as another, but haven't figured that out. Perhaps I need to solve this in the JS instead of the CSS?

resonant-gamedesign committed:
dahan growth text corrected

this adds 'to a land with Dahan' to the growth text for adding presence to a land with Dahan.

resonant-gamedesign committed:
clarified startup instructions in readme

more detailed start up instructions that help address a few hurdles i encountered

resonant-gamedesign committed:
More multi-land class updates

Perhaps this is folly. Is there a better solution?

resonant-gamedesign committed:
Revert "More multi-land class updates"

This reverts commit 69ae98b11fcb7aae72da08b5e345d5cc7a427979.

resonant-gamedesign committed:
New Growth Options (OR, AND, TEXT)

Large update to Growth options. Allows for ORs and ANDs.

resonant-gamedesign commented:
Well, I've really gotten into this. 
Big changes are primarily in the growth options. Here are some examples:
![growth_examples](https://user-images.githubusercontent.com/44596045/151743803-0fd76f65-1035-4e5a-aaba-8b227a04f2a2.PNG)

resonant-gamedesign committed:
Minor growth update to allow / in description

This update makes "Add a Presence to Land with Town or City or Blight" instead appear as "Add a Presence to Land with Town/City/Blight"

resonant-gamedesign committed:
More icon updates

All from the wiki.

resonant-gamedesign committed:
blight icon height in growth adjustment

The blight icon in growth was notably taller looking, so I scaled it down a bit.

resonant-gamedesign committed:
Growth option - adding tokens (OR and AND)

Adding growth options for adding tokens. Both AND (Many Minds) and OR (Vengeance) are implemented. Contemplating an 'INSTEAD' version but it will be far more tricky and perhaps shouldn't be handled as 'add presence'.

resonant-gamedesign committed:
updated docs for new token growths

updated the doc to provide instructions on new token growths

resonant-gamedesign committed:
Adding Coastal and Inland growth options

Also reworked Invaders using real text.

LokiMidgard commented:
> I think I'm doing something wrong. I don't mean to push all my changes to the main repository, and certainly not all at once. Any clever solutions?

create a branch form your current position. This is to not accedently loose some work
```bash
git branch safe_head
```

then reset this branch to the position you want.
Find out which commit in your history you want to revert to and copy its hash

![image](https://user-images.githubusercontent.com/389101/151823358-386c8993-a11f-45a2-8702-3fdf19dd0591.png)


```bash
git reset HASH --hard
```
Now your local branch will be several commits behind your remote.
You can override your remote with
```bash
git push --force
```

You now have your master wich is linked to this pull request at your desired position, Your newer work is safed in the branch safe_head, you can continue to work there....

I hope this helps a little bit

resonant-gamedesign commented:
Very helpful, thank you!

resonant-gamedesign committed:
Enable use of {icon} in cards

It seemed like the Card.js file wasn't calling ReplaceIcon so I added it.

resonant-gamedesign committed:
"For each" element numbers in innate descriptions

Innate powers often use a 'for each' and then a numeral in front of the elements. This update allows the proper sized numeral and font for the 'For each'. The notation is {element,#}

A few other changes snuck in. I'm working on allowing 3 element growths - that change is non-functional.

resonant-gamedesign committed:
Update board_front.js

resonant-gamedesign committed:
Corrects to Innate Power Range icons (sizing)

Improving the sizing of range icons

resonant-gamedesign committed:
Updates growth subsection line to be drawn-looking

Updates the growth subsection line to be drawn looking. Also moves the subsection lines down a bit.
Lastly, defines a class for the presence icons in the growth options.

resonant-gamedesign committed:
creates icons sized appropriately for innate notes

Innate note icons need to be smaller. This creates a space for that.

resonant-gamedesign committed:
Corrects 'move a presence' so it lines up with other growth options

Move a presence was appearing a bit 'low. This fixes that.

resonant-gamedesign committed:
Improving innate power 'notes'

resonant-gamedesign committed:
Update Misc Presence Destroy.png

removed unnecessary height

resonant-gamedesign committed:
Updates the example to use more of the new growth options

Change to highlight more growth options.

resonant-gamedesign committed:
the empty <note> object was causing spacing issues. This version never has that problem.

Removed the <note> if it is not used.

resonant-gamedesign committed:
Innate power note cleanup

tweaks to the spacing on innate notes so that the power looks good with and without a note

resonant-gamedesign committed:
Update presence track text to expand if it is long

I expanded the push/gather text to include "from 1 of your Lands', which made it longer. I needed to allow the presence text to spill over.

resonant-gamedesign committed:
Icons in threshold update (think Many Minds)

Updating the CSS to allow more icons in thresholds (like Many Minds)

resonant-gamedesign committed:
Multi-element growth improvements

Growth options - for multiple of the same element, automated the design. For OR, improved spacing.
Still need to implement AND with multiple elements.

resonant-gamedesign committed:
Enables gaining multiple different elements in growth.

This update allows the gaining of multiple different elements in growth.

resonant-gamedesign committed:
Tweaks to innate power inline element icons

Small tweaks to the inline icon placement to improve appearance.

resonant-gamedesign committed:
New badland icon, corrects color

New badland icon from the wiki

resonant-gamedesign committed:
Allows placement of more than 2 icons in a presence track

This update allows the placement of more than 2 icons in a presence track. Can handle unlimited icons, though more than 3 gets real crowded.

resonant-gamedesign committed:
Reclaim one with white background

And the accompanying css

resonant-gamedesign committed:
Update Reclaimone.png

resonant-gamedesign committed:
Adjusting town icon to remove a margin top that wasn't helping

resonant-gamedesign committed:
Final tweaks to make multiple presence track icons work

adds some css for multiple presence icons

resonant-gamedesign committed:
Icon visual tweaks

Small visual tweaks to icons in various parts of the board

resonant-gamedesign committed:
Multi Element Growth correction

Slight mistake in multi-element growth causing error for multiple of the same element. now corrected.

Yexo commented on _docs/board_front.md:
```diff
@@ -36,7 +36,14 @@ General Images: Every image in a board can be called by simply using its name en
+			- For example, add-presence(3,jungle,beast,or) as seen on Sharp Fangs or add-presence(2,town,city,blight,or) from Vengeance
+		  - add-presence(2,text,your_text_here): adds arbitrary text in case the desired option is not implemented
+		  - add-presence(2,token,and/or,token_type): allows for adding tokens in addition to or instead of presence
+		    - For example, add-presence(3,token,and,beast) is Many Minds, add-presence(1,token,or,disease) is Vengeance
```
The original file only uses spaces for indentation, but you use tabs. Most likely you don't see this in your local editor since your tab size is 4, but in the github interface with a tab size of 8 it breaks the indentation. Let's keep to one standard.

Yexo commented on _global/js/card.js:
```diff
@@ -1,5 +1,14 @@
 var quickCards = document.querySelectorAll('quick-card');
 
+window.onload = function startMain(){
```
Maybe use addEventListener like suggested here? https://stackoverflow.com/questions/16463998/multiple-javascript-window-onload-solution

That way it's possible to have an onload function in multiple script files and include all of them.

resonant-gamedesign replied:
Sounds like a good solution. I was just following along with what was done in board_front.js

Yexo replied:
Fair enough. Feel free to leave this as-is then, and after you're done I'll sent a follow-up pull request to change both instances into addEventListener.

resonant-gamedesign replied:
Great. I wasn't sure how to do what you were asking myself, which is why I didn't take it on.

Yexo commented on _global/js/card.js:
```diff
@@ -1,5 +1,14 @@
+window.onload = function startMain(){
+    const cards = document.querySelectorAll('card');
+    
+	for (i = 0; i < cards.length; ++i) {
+		var html = cards[i].innerHTML;
+		cards[i].innerHTML = replaceIcon(html);
+	}
```
This can be simplified to:

```suggestion
	for (var card of cards) {
		card.innerHTML = replaceIcon(card.innerHTML);
	}
```

Yexo commented on _global/js/general.js:
```diff
@@ -7,11 +7,22 @@ function replaceIcon(html)
   for(var match of (matchs || []))
   {
     var iconName = match.replace('{', '').replace('}', '');
+	var iconNamePieces = iconName.split(",");
```
This shows an inconsistency between spaces and tabs again. Perhaps reconfigure your editor to use spaces instead of tabs?

resonant-gamedesign replied:
I think tabs are a better convention. If we're going to set a convention going forward - shouldn't it be tabs?

Yexo commented on _global/js/general.js:
```diff
@@ -7,11 +7,22 @@ function replaceIcon(html)
     if(iconName.startsWith('no-')){
-        iconHtml = `<icon class="no"><icon class="${iconName.substring(3)}"></icon></icon>`;
+        iconHtml += `<icon class="no ${iconName.substring(3)}"></icon>`;
+/*         iconHtml = `<icon class="no"><icon class="${iconName.substring(3)}"></icon></icon>`; */
```
```suggestion
```

Yexo commented on _examples/card/raw-html-card-front.html:
```diff
@@ -2,6 +2,8 @@
   <link href="../../_global/css/global.css" rel="stylesheet" />
   <link href="../../_global/css/card.css" rel="stylesheet" />
+  <script type="text/javascript" src="../../_global/js/general.js"></script>
+  <script type="text/javascript" src="../../_global/js/card.js"></script>  
```
Are these additional includes required?

resonant-gamedesign replied:
the {icon} shortcuts were not working until I added these

Yexo commented on _global/js/board_front.js:
```diff
@@ -415,14 +598,27 @@ function getPresenceNodeHtml(nodeText, first, trackType, addEnergyRing) {
-            subText = Capitalise(splitOptions[0])+", "+subText;
-
-            var top = "";
+            subText = Capitalise(splitOptions[0])+", "+subText; */
```
Please actually remove code instead of only commenting it out. The old code can always be found back in git history (that's one of the reasons to use version control in the first place).

resonant-gamedesign replied:
Yup, makes sense. Learning the etiquette as I go. I thought that the old code would help others see the changes - but Git has me covered there too.

Yexo commented on _global/js/board_front.js:
```diff
@@ -435,7 +631,30 @@ function getPresenceNodeHtml(nodeText, first, trackType, addEnergyRing) {
 
             var inner = "<icon-top>"+top+"</icon-top>" +
-                "<icon-bottom>"+bottom+"<icon-bottom>";
+                "<icon-bottom>"+bottom+"<icon-bottom>"; */
```
Same here, please actually remove the old code.

Yexo reviewed (commented):
Overall looks good, I've left a bunch of comments. Please take a look, most should be easy to fix

resonant-gamedesign committed:
Updating to align with separate parseInnatePower function

This change aligns my code with the 'parseInnatePower' function that was added to allow Aspects.

Also captured some other cleanup.

resonant-gamedesign committed:
Minor cleanup on card.js

slightly more efficient code

resonant-gamedesign commented:
I believe the conflict that Git is identifying is not really a conflict. We can just go with my version of the code, which I updated to use the parseInnatePower and parseInnatePowers functions added by the Aspect update.

But my only option is to COMMIT after I resolve the conflict. Is that OK?

resonant-gamedesign committed:
Correcting tabs vs spaces

Updating docs to use spaces intsead of tabs. May take time to catch all the errors.

resonant-gamedesign committed:
Update _global/js/general.js

Co-authored-by: Thijs Marinussen <t.marinussen@gmail.com>

Yexo commented:
> I believe the conflict that Git is identifying is not really a conflict. We can just go with my version of the code, which I updated to use the parseInnatePower and parseInnatePowers functions added by the Aspect update.
> 
> But my only option is to COMMIT after I resolve the conflict. Is that OK?

Yes, that's normal and expected. Git calls it a conflict because multiple branches modified the same lines (or lines very close to eachter), so git doesn't know how to automatically resolve it. A "conflict" just means you manually have to verify or update the code, the "commit" then tells git "I verified/updated the code and now it's ok".

resonant-gamedesign committed:
use the higher resolution 'fast' icon

no reason to use teh 25px version when the 50px is already in the file.

resonant-gamedesign committed:
updates to fast and slow icons in special rules

resonant-gamedesign committed:
Merge branch 'master' of https://github.com/resonant-gamedesign/spirit-island-template

resonant-gamedesign committed:
Growth option: Gain X per Y

Adds a few modes to the growth option for adding X energy per Y element, as seen on Wildfire. Allows the exact style shown on Wildfire (ie. +2 then +1 per Fire) or simply +1 per Element (not seen on any official spirit).

Also, some updates to tabs.

resonant-gamedesign committed:
Tweaks to reclaim to make it work in growth

resonant-gamedesign committed:
Tweaks to css icons in the info-targeting for innates

Some re-naming makes the icon specifications more efficient (using innate-info instead of innate-info-target, for example)

resonant-gamedesign committed:
Small tweaks to icons in innate descriptions

resonant-gamedesign committed:
More small icon tweaks

resonant-gamedesign committed:
Improve the range icon

resonant-gamedesign committed:
allows the user to do a bullet in the special rule

but not the best solution

resonant-gamedesign reviewed (commented):
An example of new growth options
![Capture_updates](https://user-images.githubusercontent.com/44596045/152929931-ee3d55cf-09c5-45cb-a7c2-03ee65b798e4.PNG)

resonant-gamedesign committed:
Merge branch 'master' into master

resonant-gamedesign commented:
Hey folks! I understand that this action can be reversed, so I'm going to go ahead and merge my updates. I don't think anyone is going to take the time to review them and I'm confident that it all works. There are a number of big improvements that I think benefit the project, even if minor issues are introduced (which we can resolve in future edits).

The advantage of merging this is that it allows me to continue developing and perhaps try to merge the other changes folks have proposed.
